include/ay8910.h

```diff
@@ -0,0 +1,125 @@
+#ifndef __LIBDRAGON_AY8910_H
+#define __LIBDRAGON_AY8910_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief Decimation factor for AY8910. 
+ * 
+ * AY8910 is usually clocked at a very high frequency (>100K), and thus it
+ * requires downsampling to be played back. This emulator offers a very basic
+ * downsampling filter via decimation (taking average of Nth consecutive samples)
+ * that is usually a good compromise between quality and speed, for realtime
+ * playback. It will not sound as good as a real downsampling filter though.
+ * 
+ * It is suggested to configure this number to the smallest value that brings
+ * the AY8910 output frequency within the playback sample rate.
+ */
+#define AY8910_DECIMATE           3
+
+/** @brief Generate stereo output.
+ * 
+ * If 1, AY8910 will generate a stereo output with fixed pans for
+ * each of the three internal channels, similar to what Arkos Tracker 2 does
+ * in stereo mode.
+ */
+#define AY8910_OUTPUT_STEREO      1
+
+/** @brief Define the global attenuation applied to volumes (range 0.0 - 1.0). 
+ * 
+ * The AY8910 often clips so it's important to lower a bit the volume to avoid
+ * sound artifacts.
+ */
+#define AY8910_VOLUME_ATTENUATE   0.8
+
+/** @brief Generate silence as 0.
+ * 
+ * Normally, AY8910 generates output samples in which the silence is
+ * represented by -32768 (minimum volume). This is a little inconvenient
+ * if the caller wants to skip generation when the AY8910 is muted for
+ * performance reasons, because audio mixers normally assume that muted
+ * channels are made by samples with value 0, otherwise disabling the
+ * AY8910 would affect the volume of all other channels.
```

Comment:
Also I'm not sure I follow that conversion how does a 4bit value of 0 becomes -2^15?

Reply:
> having a constant value on one of the channels should not affect the volume

I can check that again. This code predates the mixer BTW.

> It would introduce a DC bias though, which would decrease the available headroom. i.e it would be easier to clip.

That for sure is a problem in itself. 

> How much it matters relates to the scaling of the output.

I've added a macro to lower the volume of the AY-3-8910 specifically for this. It's not that useful when you activate the center volume feature, in fact.

> Another thing to keep in mind, a channel with a non zero value can create a pop initially

Yes that happened pre-mixer, and I also had a comment here about that. It doesn't happen now because the mixer does a volume ramp in RSP specifically to avoid this (both when the volume suddenly changes during playback, and for start/stop).

> Also I'm not sure I follow that conversion how does a 4bit value of 0 becomes -2^15?

How do you convert an unsigned 4 bit amplitude into a 16 bit signed amplitude? The normal formula would be to map 0 to -2^15, and 15 to 2^15, with the other values in-between.

Reply:
What I thought initially was mapping 0-15 to 0-2^15 and I think this is what the define already does. From the descrption I thought it was causing a drop on representable sample resolution but thats not true as I now better understand it. I wonder where the original chip's output offset is. Then I suspect the "dullness" is coming from the fact that we are using a sort of "rectified" output which never crosses the outputs center point. Then the ideal solution would be adding a DC removal filter but that might be an overkill.